CONTRIBUTING.md

```diff
@@ -0,0 +1,303 @@
+on your system. _While Nix is also capable of building Nickel by itself, using
+cargo (either a system-wide installation or the one given by the Nix development
+shell) is the recommended way of building when working on the Nickel repository
+itself_. The reason is that incremental compilation for Rust and Nix is not
+there yet, and incremental rebuilds using only Nix are going to be painfully
+long.
+
+## Content
+
+The Nickel repository consist in 3 crates:
+
+- `nickel-lang` (path: `.`). The main crate containing the interpreter as a library as well as the `nickel` binary.
+- `nickel-lang-lsp` (path: `lsp/nls/`). the Nickel Language Server (NLS), an LSP server for Nickel.
+- `nickel-lang-utilities`: (path: `utilities/`). An auxiliary crate regrouping
+   helpers for tests and benchmarks. Not required to build `nickel` itself.
+
+Other noteworthy items:
+
+- The user manual in `doc/manual/`, as a bunch of markdown files.
+- A VSCode extension for NLS in `lsp/client-extension/`.
+
+## Using Nix
+
+To set up a development environment using a recent Nix (>= 2.4):
+
+1. Clone the repository: `git clone git@github.com:tweag/nickel.git`
+2. At the root of the repository, run `nix develop`. You should now be dropped
+   in a shell with all the required tool to hack on Nickel (`rust`, `cargo`,
+   etc.)
+
+## Without Nix
+
+Otherwise, you can install the Rust toolchain in a standard way:
+
+1. [Install Rust][install-rust]
+2. [Install Cargo][install-cargo]
+
+## Building
+
+### Nickel
+
+To build the main crate, just run `cargo build` at the root of this repository.
+Upon success, the binary will be available somewhere inside the `target`
+directory.
+
+You can make the binary available in your path via `cargo install` at the root
+of this repository:
+
+```shell
+$ cargo install --path .
+$ nickel --version
+nickel-lang 0.1.0
+```
+
+### NLS (nickel-lang-lsp)
+
+To build NLS, the LSP server for Nickel, build the crate inside `lsp/nls`:
+
+```shell
+$ cd lsp/nls
+lsp/nls $ cargo build
+```
+
+As for Nickel, you can make the `nls` binary available in your path via `cargo
+install`:
+
+```shell
+$ cargo install --path .
+$ nls --version
+nickel-lang-lsp 0.1.0
+```
+
+### WebAssembly REPL
+
+This repository is able to build WebAssembly (WASM) version of the REPL, which
+is used for the online playground on [nickel-lang.org][nickel-lang.org]. While
+using `cargo` directly is the recommended way for building, both methods have
+advantages for the WASM REPL. Building through Cargo alone is possible, but
+requires installing new tools and patching the `Cargo.toml`. On the other hand,
+`Nix` can perform the build in one simple command, but incremental compilation
+is not as good as with direct usage of `cargo`.
+
+Both methods are described below.
+
+#### Using Nix
+
+At the root of the repository, run:
+
+```shell
+$ nix build .#buildWasm
+```
+
+The corresponding NPM package is put in `result/`.
+
+#### Using Cargo
```

Comment:
These instructions are horrible ([remember](https://discourse.tweag.io/t/make-your-instructions-snap/157)). Is there really no better way?

Reply:
It's fundamentally annoying, because of this patching Cargo.toml. I guess we could just drop one `sed` invocation that do steps 2 and 3., which is what the flake is doing during the build phase. I'll try to make it better.

Reply:
I want to express my puzzlement at the fact that we need to patch the `Cargo.toml`. Why do we need to do something like that?

Reply:
Let me reproduce the comments inside the corresponding build phase in our `flake.nix`:

```
# Wasm-pack requires to change the crate type. Cargo doesn't yet
# support having different crate types depending on the target, so
# we switch there
sed -i 's/\[lib\]/[lib]\ncrate-type = ["cdylib", "rlib"]/' Cargo.toml
                                                                                
# This is a hack to prevent the fs2 crate from being compiled on wasm.
# This may be able to be removed once one or more of these issues are resolved:
# https://github.com/bheisler/criterion.rs/issues/461
# https://github.com/rust-lang/cargo/issues/1596
# https://github.com/rust-lang/cargo/issues/1197
# https://github.com/rust-lang/cargo/issues/5777
sed -i '/nickel-lang-utilities/d' Cargo.toml
```

The related issue for the first problem is e.g. https://github.com/rust-lang/cargo/issues/4881. May be useful to add it in the comments, by the way.

Reply:
Building the crate in "normal" mode and for WASM requires two different incompatible configurations (the crate type), and Cargo doesn't allow to have this configuration depends on the target currently.

For the second patch, it's related to `utilities` not being used for the WASM build whatsoever (it's just a bunch of helpers for tests and benchmarks), but cargo still insist on building it in WASM because it appears in the dependencies. But `utilities` has some filesystem related helpers that doesn't please the WASM build.

Reply:
How about maintaining to `cargo.toml` files? (the second one is called, say `cargo.wasm.yaml`) Would that be possible?

Reply:
A third possibility is to have a separate branch `wasm`, that follows `closely` with the patched manifest. We'll have to keep the branch in sync, but that may be the less annoying solution for users.

Reply:
No, a separate branch sounds awful. Having two nearly identical cargo file sounds merely annoying (it can be solved easily after CI). But Cargo not letting me switch between files basically kills this idea (we can make two different cargo files, and use a wrapper around cargo which symlinks whichever file is relevant for the build; this is terrible, let's not).

The shell script is not a great solution. Not even a good solution. But it's not a bad solution either. So we should probably go with that. It still runs the risk of people committing the wasm `cargo.yaml` by mistakes, though. But it's not too hard to deal with.

What do other project which compile to wasm do? Do they patch the `cargo.toml` file as well?

Reply:
After digging around a bit, I find a lot of people complaining but not many better solutions. There is even a cargo plugin to do the dirty patching automatically: https://crates.io/crates/cargo-crate-type. In some situation you can specify both crate types `["cdylib", "rlib"]`, but now I remember that I tried this trick and it broke macOS build because something is not fully supported there. I gave a look at some repositories that compile to WASM but it's not always clear to see what's happening on big multi-crate repo to be honest. Some use other build system on top of Cargo (NPM, thunk, Makefile), some use emscripten instead of wasm-pack as a WASM toolchain. I suspect some of them just have a separate crate with additional code for the WASM version anyway, beside the `crate-type` difference.

One technique mentioned is to have a proxy crate, say `nickel-lang-wasm` that is just wrapping `nickel-lang` as a dependency and change the crate type to `cdylib`. It's possible that it'll also solve the problem of `utilities` being compiled, because it's a dev-dependency and I guess that shouldn't propagate to the wrapping crate. That may be a less awful solution to try.

Reply:
> One technique mentioned is to have a proxy crate, say `nickel-lang-wasm` that is just wrapping `nickel-lang` as a dependency and change the crate type to `cdylib`. It's possible that it'll also solve the problem of `utilities` being compiled, because it's a dev-dependency and I guess that shouldn't propagate to the wrapping crate. That may be a less awful solution to try.

This very much sounds like progress, actually. Let me know whether this works. (I'll look at the rest of your changes independently, no need to hold this PR for that. But I'm happy to see that we have potential solutions).